File: Go/ParkingLot2-go/vehicle.go
package main

import "fmt"

type Vehicle struct {
	Type   string
	Number string
	Size   string
}

func NewBike(number, size string) *Vehicle {
	return &Vehicle{Type: "Bike", Number: number, Size: size}
}

func NewCar(number, size string) *Vehicle {
	return &Vehicle{Type: "Car", Number: number, Size: size}
}

func NewTruck(number, size string) *Vehicle {
	return &Vehicle{Type: "Truck", Number: number, Size: size}
}

func CreateVehicle(vehicleType, number string) (*Vehicle, error) {
	if vehicleType == "" || number == "" {
		return nil, fmt.Errorf("invalid vehicle inputs!")
	}
	switch vehicleType {
	case "bike", "Bike":
		return NewBike(number, "small"), nil
	case "car", "Car":
		return NewCar(number, "medium"), nil
	case "truck", "Truck":
		return NewTruck(number, "large"), nil
	default:
		return nil, fmt.Errorf("unsupported vehicle type: %s", vehicleType)
	}
}

File: JavaScript/ParkingLot2/Vehicle.js
class Vehicle {
  type = '';
  number = '';
  size = '';

  constructor(type, number, size) {
    if (!type || !number || !size) {
      throw new Error('Invalid vehicle!');
    }
    this.type = type;
    this.size = size;
    this.number = number;
  }
}

class Bike extends Vehicle {
  constructor(number, size) {
    super('Bike', number, size);
  }
}

class Car extends Vehicle {
  constructor(number, size) {
    super('Car', number, size);
  }
}

class Truck extends Vehicle {
  constructor(number, size) {
    super('Truck', number, size);
  }
}

class VehicleFactory {
  static create(type, number) {
    if (!type || !number) {
      throw new Error('Invalid vehicle inputs!');
    }

    const normalized = type.toLowerCase();

    switch (normalized) {
      case 'bike':  return new Bike(number, 'small');
      case 'car':   return new Car(number, 'medium');
      case 'truck': return new Truck(number, 'large');

      default:
        throw new Error(`Unsupported vehicle type: ${type}`);
    }
  }
}

module.exports = { VehicleFactory };
